processor/metricsgenerationprocessor/utils.go

```diff
@@ -50,36 +51,140 @@ func getMetricValue(metric pmetric.Metric) float64 {
-				case pmetric.MetricTypeGauge:
-					dataPointCount = newMetric.Gauge().DataPoints().Len()
-				}
+				newMetric := generateMetricFromMatchingAttributes(metric, metric2, rule, logger)
+				appendNewMetric(ilm, newMetric, rule.name, rule.unit)
+			}
+		}
+	}
+}
+
+// Calculates a new metric based on the calculation-type rule specified. New data points will be generated for each
+// calculation of the input metrics where overlapping attributes have matching values.
+func generateMetricFromMatchingAttributes(metric1 pmetric.Metric, metric2 pmetric.Metric, rule internalRule, logger *zap.Logger) pmetric.Metric {
+	var metric1DataPoints pmetric.NumberDataPointSlice
+	var toDataPoints pmetric.NumberDataPointSlice
+	to := pmetric.NewMetric()
 
-				// Only create a new metric if valid data points were calculated successfully
-				if dataPointCount > 0 {
-					appendMetric(ilm, newMetric, rule.name, rule.unit)
+	// Setup to metric and get metric1 data points
+	switch metricType := metric1.Type(); metricType {
+	case pmetric.MetricTypeGauge:
+		to.SetEmptyGauge()
+		metric1DataPoints = metric1.Gauge().DataPoints()
+		toDataPoints = to.Gauge().DataPoints()
+	case pmetric.MetricTypeSum:
+		to.SetEmptySum()
+		metric1DataPoints = metric1.Sum().DataPoints()
+		toDataPoints = to.Sum().DataPoints()
+	default:
+		logger.Debug(fmt.Sprintf("Calculations are only supported on gauge or sum metric types. Given metric '%s' is of type `%s`", metric1.Name(), metricType.String()))
+		return pmetric.NewMetric()
+	}
+
+	// Get metric2 data points
+	var metric2DataPoints pmetric.NumberDataPointSlice
+	switch metricType := metric2.Type(); metricType {
+	case pmetric.MetricTypeGauge:
+		metric2DataPoints = metric2.Gauge().DataPoints()
+	case pmetric.MetricTypeSum:
+		metric2DataPoints = metric2.Sum().DataPoints()
+	default:
+		logger.Debug(fmt.Sprintf("Calculations are only supported on gauge or sum metric types. Given metric '%s' is of type `%s`", metric2.Name(), metricType.String()))
+		return pmetric.NewMetric()
+	}
+
+	for i := 0; i < metric1DataPoints.Len(); i++ {
+		metric1DP := metric1DataPoints.At(i)
+
+		for j := 0; j < metric2DataPoints.Len(); j++ {
+			metric2DP := metric2DataPoints.At(j)
+			if dataPointAttributesMatch(metric1DP, metric2DP) {
+				val, err := calculateValue(dataPointValue(metric1DP), dataPointValue(metric2DP), rule.operation, rule.name)
+
+				if err != nil {
+					logger.Debug(err.Error())
+				} else {
+					newDP := toDataPoints.AppendEmpty()
+					metric1DP.CopyTo(newDP)
+					newDP.SetDoubleValue(val)
+
+					metric2DP.Attributes().Range(func(k string, v pcommon.Value) bool {
+						switch v.Type() {
+						case pcommon.ValueTypeDouble:
+							newDP.Attributes().PutDouble(k, v.Double())
+
+						case pcommon.ValueTypeInt:
+							newDP.Attributes().PutInt(k, v.Int())
+						case pcommon.ValueTypeBool:
+							newDP.Attributes().PutBool(k, v.Bool())
+						default:
+							newDP.Attributes().PutStr(k, v.Str())
+						}
```

Comment:
No need to track all possible types
```suggestion
						v.CopyTo(newDP.Attributes().PutEmpty(k))
```

Reply:
Thanks Dmitrii, that's much better 👍